File: HPWorkpath_v1.6.3/Samples/ExampleExtensions/source/GoogleSigninSample/src/com/hp/workpath/sample/googlesigninsample/task/RevokeTokenTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.googlesigninsample.task;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.webkit.CookieManager;

import androidx.annotation.NonNull;

import com.hp.workpath.sample.googlesigninsample.MainActivity;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class RevokeTokenTask {

    private static final String TAG = MainActivity.TAG;

    private String mAccessToken;
    private TaskInterface mTaskInterface;
    private Throwable mThrowable;
    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public RevokeTokenTask(@NonNull String accessToken, @NonNull TaskInterface taskInterface) {
        this.mAccessToken = accessToken;
        this.mTaskInterface = taskInterface;
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        OkHttpClient client = new OkHttpClient().newBuilder()
                                .connectTimeout(60, TimeUnit.SECONDS)
                                .readTimeout(60, TimeUnit.SECONDS)
                                .writeTimeout(60, TimeUnit.SECONDS)
                                .build();

                        Log.i(TAG, "Revoke accessToken: " + mAccessToken);
                        Request request = new Request.Builder()
                                .url("https://accounts.google.com/o/oauth2/revoke?token=" + mAccessToken)
                                .method("POST", RequestBody.create(null, new byte[0]))
                                .build();
                        Response response = client.newCall(request).execute();
                        if (response != null) {
                            if (response.isSuccessful()) {
                                onPostExecute(true);
                            } else {
                                mThrowable = new Exception(response.body().string());
                                onPostExecute(false);
                            }
                        } else {
                            mThrowable = new Exception("Revoke token response is null");
                            onPostExecute(false);
                        }
                    } catch (Throwable t) {
                        mThrowable = t;
                        onPostExecute(false);
                        executor.shutdown();
                    }
                }
            });
        } catch (Throwable t) {
            mThrowable = t;
            onPostExecute(false);
            executor.shutdown();
        }
    }

    protected void onPostExecute(boolean result) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (!result) {
                    mTaskInterface.onFailure(mThrowable);
                } else {
                    mTaskInterface.revokedToken();
                }
                CookieManager.getInstance().removeAllCookie();
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleExtensions/source/GoogleSigninSample/src/com/hp/workpath/sample/googlesigninsample/task/RevokeTokenTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.googlesigninsample.task

import android.util.Log
import android.webkit.CookieManager
import com.hp.workpath.sample.googlesigninsample.MainActivity
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

class RevokeTokenTask(private val mAccessToken: String, private val mTaskInterface: TaskInterface) {
    private var mThrowable: Throwable? = null

    suspend fun execute() {
        var result = false
        try {
            val client = OkHttpClient().newBuilder()
                    .connectTimeout(60, TimeUnit.SECONDS)
                    .readTimeout(60, TimeUnit.SECONDS)
                    .writeTimeout(60, TimeUnit.SECONDS)
                    .build()
            Log.i(TAG, "Revoke accessToken: $mAccessToken")
            val request = Request.Builder()
                    .url("https://accounts.google.com/o/oauth2/revoke?token=$mAccessToken")
                    .method("POST", RequestBody.create(null, ByteArray(0)))
                    .build()
            val response = client.newCall(request).execute()
            if (response != null) {
                if (response.isSuccessful) {
                    result = true
                } else {
                    Throwable(response.body()?.string())
                }
            } else {
                Throwable("Revoke token response is null")
            }
        } catch (e: IOException) {
            mThrowable = e
        }
        onPostExecute(result)
    }

    private suspend fun onPostExecute(result: Boolean) {
        if (!result) {
            mTaskInterface.onFailure(mThrowable)
        } else {
            mTaskInterface.revokedToken()
        }
        CookieManager.getInstance().removeAllCookies(null)
    }

    companion object {
        private val TAG: String = MainActivity.TAG
    }
}